nixos/modules/services/x11/extra-layouts.nix

```diff
@@ -0,0 +1,154 @@
+        description = "A short description of the layout.";
+      };
+
+      languages = mkOption {
+        type = types.listOf types.str;
+        description =
+        ''
+          A list of languages provided by the layout.
+          (Use ISO 639-2 codes, for example: "eng" for english)
+        '';
+      };
+
+      filepath = mkOption {
+        type = types.path;
+        description = "The path to the layout file.";
+      };
+    };
+  };
+
+  patchIn = name: layout:
+  ''
+      # install layout file
+      cp ${layout.filepath} symbols/${name}
+
+      # patch makefiles
+      ${pkgs.ed}/bin/ed -v symbols/Makefile.am <<EOF
+      /symbols_DATA =
+      a
+      ${name} \\
+      .
+      w
+      EOF
+      ${pkgs.ed}/bin/ed -v symbols/Makefile.in <<EOF
+      /symbols_DATA =
+      a
+      ${name} \\
+      .
+      w
+      EOF
+
+      # add layout description
+      ${pkgs.ed}/bin/ed -v rules/base.xml.in <<EOF
+      /<\/layoutList>
+      -
+      a
+      <layout>
+        <configItem>
+          <name>${name}</name>
+          <_shortDescription>${name}</_shortDescription>
+          <_description>${layout.description}</_description>
+          <languageList>
+            ${lib.concatMapStrings (lang: "<iso639Id>${lang}</iso639Id>\n") layout.languages}
+          </languageList>
+        </configItem>
+        <variantList/>
+      </layout>
+      .
+      w
+      EOF
+      cp rules/base.xml.in rules/base.evdev.xml.in
+  '';
+
+  xkb_patched = pkgs.xorg.xkeyboardconfig.overrideAttrs (old: {
+    buildInputs = old.buildInputs ++ [ pkgs.automake ];
+    postPatch = with lib; concatStrings (mapAttrsToList patchIn layouts);
+  });
+
+
+in
+
+{
+
+  ###### interface
+
+  options.services.xserver = {
+    extraLayouts = mkOption {
+      type = types.attrsOf (types.submodule layoutOpts);
+      default = {};
+      example = literalExample
+      ''
+        {
+          mine = {
+            description = "My custom xkb layout.";
+            languages = [ "eng" ];
+            filepath = /path/to/my/layout;
+          };
+        }
+      '';
+      description = ''
+        Extra custom layouts that will be included in the xkb configuration.
+        Information on how to create a new layout can be found here:
+        https://www.x.org/releases/current/doc/xorg-docs/input/XKB-Enhancing.html#Defining_New_Layouts
+      '';
+    };
+
+  };
+
+  ###### implementation
+
+  config = mkIf (layouts != { }) {
+
+    # We don't override xkeyboard_config directly to
+    # reduce the amount of packages to be recompiled.
+    # Only the following packages are necesary to set
+    # a custom layout anyway:
+    nixpkgs.overlays = lib.singleton (self: super: {
+      xorg = super.xorg // {
+        xorgserver = super.xorg.xorgserver.overrideAttrs (old: {
+          configureFlags = old.configureFlags ++ [
+            "--with-xkb-bin-directory=${self.xorg.xkbcomp}/bin"
+            "--with-xkb-path=${xkb_patched}/share/X11/xkb"
```

Comment:
Oh actually, doesn't the `services.xserver.xkbDir` option do the same, but without having to recompile?

Reply:
Yeah, but it doesn't really work: see #31138.